subsystems/mvdm/ntvdm/bios/vidbios.c

```diff
@@ -2942,6 +2942,303 @@ static VOID VidBiosPrintCharacter(CHAR Character, BYTE Attribute, BOOLEAN UseAtt
+
+            EmulatorReadMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            if (Xor)
+            {
+                MemoryValue ^= Value << ValueShift;
+            }
+            else
+            {
+                MemoryValue &= ~(MaxColor << ValueShift);
+                MemoryValue |= Value << ValueShift;
+            }
+            EmulatorWriteMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            break;
+        }
+
+        case 0x0D: /* EGA 320x200 16c */
+        case 0x0E: /* EGA 640x200 16c */
+        case 0x10: /* EGA 640x350 16c */
+        case 0x12: /* VGA 640x480 16c */
+        {
+            ULONG Address;
+            BOOL Xor;
+            BYTE OldMode;
+            BYTE OldRotation;
+            BYTE OldRead;
+            BYTE OldMask;
+            BYTE CurrentPlaneMask;
+            BYTE CurrentPlane;
+            BYTE Shift = ~X & 7;
+
+            Xor = (Value & 0x80) != 0;
+            Value &= 15;
+            Address = TO_LINEAR(GRAPHICS_VIDEO_SEG,
+                                Page * Bda->VideoPageSize + (Y * ScreenWidth + X) / 8);
+
+            /* Make sure we're in write mode 0 */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_MODE_REG);
+            OldMode = IOReadB(VGA_GC_DATA);
+            IOWriteB(VGA_GC_DATA, 0x00);
+
+            /* Setup the rotation */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_ROTATE_REG);
+            OldRotation = IOReadB(VGA_GC_DATA);
+            IOWriteB(VGA_GC_DATA, Xor ? 0x18 : 0x00);
+
+            /* Save old read */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_READ_MAP_SEL_REG);
+            OldRead = IOReadB(VGA_GC_DATA);
+
+            /* Save old mask */
+            IOWriteB(VGA_SEQ_INDEX, VGA_SEQ_MASK_REG);
+            OldMask = IOReadB(VGA_SEQ_DATA);
+
+            /* Loop and write over all four color bitplanes */
+            CurrentPlaneMask = 1;
+            for (CurrentPlane = 0; CurrentPlane < 4; CurrentPlane++)
+            {
+                BYTE Dummy;
+                BYTE ActualValue;
+
+                IOWriteB(VGA_GC_INDEX, VGA_GC_READ_MAP_SEL_REG);
+                IOWriteB(VGA_GC_DATA, CurrentPlane);
+                IOWriteB(VGA_SEQ_DATA, CurrentPlaneMask);
+                EmulatorReadMemory(&EmulatorContext, Address, &Dummy, sizeof(BYTE));
+
+                /* There's probably a way to handle this logic on write, but I'll just leave it at that */
+                if (Xor)
+                {
+                    ActualValue = (Value & 1) << Shift;
+                }
+                else
+                {
+                    ActualValue = Dummy;
+                    ActualValue &= ~(1 << Shift);
+                    ActualValue |= (Value & 1) << Shift;
+                }
+
+                EmulatorWriteMemory(&EmulatorContext, Address, &ActualValue, sizeof(BYTE));
+
+                Value >>= 1;
+                CurrentPlaneMask <<= 1;
+            }
```

Comment:
```suggestion

            /* Write to all planes */
            IoWriteB(VGA_SEQ_DATA, 0x0F);

            /* Set the bit mask to only include one pixel */
            IoWriteB(VGA_GC_INDEX, VGA_GC_BITMASK_REG);
            OldBitMask = IoReadB(VGA_GC_DATA);
            IoWriteB(VGA_GC_DATA, 1 << ((Y * ScreenWidth + X) & 0x07));

            BYTE Dummy;
            /* Do a read just to load the latch register. */
            EmulatorReadMemory(&EmulatorContext, Address, &Dummy, sizeof(BYTE));

            /* In write mode 2, this value will be expanded across all 8 bits, but the bit mask will protect the other pixels */
            EmulatorWriteMemory(&EmulatorContext, Address, &Value, sizeof(BYTE));
            
            /* Restore the bit mask */
            IoWriteB(VGA_GC_INDEX, VGA_GC_BITMASK_REG);
            IoWriteB(VGA_GC_DATA, OldBitMask);

```

Reply:
Please excuse the possible (and probable) errors in there, as I only wrote it from memory without even checking if it compiles. But hopefully it will help you figure out the logic behind doing this with VGA in a fast way, that you can also then apply to the other modes.